### PR TITLE
Chebyshev polynomial basis implementation and test

### LINKED ISSUE #10: Task 6
Melissa Wang, Chebyshev

### PR DESCRIPTION
**Pull request by Lemux133**
## Description
This module introduces Chebyshev polynomial basis (task 6). It provides a class `ChebyshevBasis` that inherits from `PolynomialBasis` and implements:

- **evaluate_basis**: evaluates individual Chebyshev polynomials T_n(x) = cos(n*arccos x)  on [-1, 1]

- **fit**: computes interpolation coefficients by building the Vandermonde matrix V_{ij} = T_j(x_i) and solving Vc = y

- **evaluate**: evaluates the full Chebyshev series \sum c_i T_i(x)

- **chebyshev_nodes**: generates optimal interpolation points on [-1, 1] for both roots and extrema

## Linked Issue
Fixes #10

## Implementation Checklist
- [x] My class inherits from `PolynomialBasis` (if applicable).
- [x] I have implemented all abstract methods (`evaluate_basis`, `fit`, etc.).
- [x] I have used `numpy` vectorization to avoid inefficient Python loops.
- [x] Documentation strings (docstrings) are provided for all public methods.

### 1. Mathematical Integrity
- [ ] **Convergence:** Does the implementation maintain expected accuracy? (Checked for Runge's phenomenon or stability).
- [ ] **Edge Cases:** Does it handle single points or non-equally spaced nodes correctly?
- [ ] **Domain Mapping:** Does the class correctly map the input domain to the basis domain (e.g., [-1, 1] for Chebyshev)?

### 2. API & Style
- [ ] **Method Signatures:** Methods match the Abstract Base Class exactly.
- [ ] **Naming:** Follows `CamelCase` for classes and `snake_case` for functions/variables.
- [ ] **Clean Code:** No "magic numbers"; constants are named and explained.

### 3. Testing Quality
- [ ] **Coverage:** Unit tests cover typical usage and edge cases.
- [ ] **Precision:** Tests use `np.testing.assert_allclose` instead of `==`.
- [ ] **Validation:** There is a test proving the basis can exactly reconstruct a polynomial of its own degree.

## Screenshots / Plots

### CODE REVIEW

File: src/numanalysislib/basis/chebyshev.py [hunk 1]
```diff
@@ -0,0 +1,187 @@
...
+    def __init__(self, degree: int) -> None:
+        """
+        initialize the Chebyshev basis on [-1, 1]
+
+        parameters
+        degree: int; maximum polynomial deg
+        """
+        super().__init__(degree, a=-1.0, b=1.0)
+        # for instance, if we pass degree=5 when creating a ChebyshevBasis object,
+        # then the PolynomialBasis constructor would set self.degree=3, self.n_dofs=4
```
Comment (rleburu) [inline src/numanalysislib/basis/chebyshev.py]: This comment seems incorrect (or outdated), passing degree=5 should give self.degree=5 and self.n_dofs=6. Might want to fix or remove to avoid confusion.

File: src/numanalysislib/basis/chebyshev.py [hunk 2]
```diff
@@ -0,0 +1,187 @@
...
+
+        # warning for high-degree evaluation which may be unstable
+        if ind > 50:
+            warnings.warn(
+                f"evaluating high-degree Chebyshev polynomial T_{ind}(x) "
+                f"using cos(n*arccos(x)) may be numerically unstable",
+                RuntimeWarning
+            )
+
+        return np.cos(ind*np.arccos(x))
```
Comment (rleburu) [inline src/numanalysislib/basis/chebyshev.py:64]: The current implementation is correct, but as you mentioned and addressed, cos(n*arccos(x)) can be unstable for large degrees near the endpoints. 

We could consider using the three-term recurrence (the formula is on page 428 of the Quarteroni book :) ), which should be more stable.

**Review by rleburu — COMMENTED** [review]
Great work! The implementation is nice, and the tests are thorough. No major issues I caught. The comments I've left are minor suggestions and nitpicks.

IMPORTANT NOTE: It looks like a couple of integrator files (integrator.py and test_integrator.py) got included in this PR, you might want to clean up the branch so the PR only contains the Chebyshev files. (I don't think this was intentional :), but let me know it was, as I didn't review these files)

## Peer Review Checklist (To be completed by Reviewer --- Rishi)

### 1. Mathematical Integrity
- [x] **Convergence:** Does the implementation maintain expected accuracy? (Checked for Runge's phenomenon or stability).
- [x] **Edge Cases:** Does it handle single points or non-equally spaced nodes correctly?
- [x] **Domain Mapping:** Does the class correctly map the input domain to the basis domain (e.g., [-1, 1] for Chebyshev)?

### 2. API & Style
- [x] **Method Signatures:** Methods match the Abstract Base Class exactly.
- [x] **Naming:** Follows `CamelCase` for classes and `snake_case` for functions/variables.
- [x] **Clean Code:** No "magic numbers"; constants are named and explained.

### 3. Testing Quality
- [x] **Coverage:** Unit tests cover typical usage and edge cases.
- [x] **Precision:** Tests use `np.testing.assert_allclose` instead of `==`.
- [x] **Validation:** There is a test proving the basis can exactly reconstruct a polynomial of its own degree.

## Screenshots / Plots

File: src/numanalysislib/basis/chebyshev.py (hunk 2 above)
Comment (Lemux133) [reply src/numanalysislib/basis/chebyshev.py:64]: Thanks for the info!!

File: src/numanalysislib/basis/chebyshev.py [hunk 3]
```diff
@@ -0,0 +1,187 @@
...
+    parameters
+    1. degree: int; the maximum degree of the polynomial basis
+    """
+
+    def __init__(self, degree: int) -> None:
+        """
+        initialize the Chebyshev basis on [-1, 1]
+
+        parameters
+        degree: int; maximum polynomial deg
```
Comment (mtezzele) [inline src/numanalysislib/basis/chebyshev.py]: please use the proper doc style (check if it is rendered correctly locally).

File: src/numanalysislib/basis/chebyshev.py [hunk 4]
```diff
@@ -0,0 +1,187 @@
...
+
+        raises ValueError
+               if index is out of range for the basis degree
+        """
+        if ind < 0 or ind > self.degree:
+            raise ValueError(
+                f"basis index {ind} is out of range for degree {self.degree}"
+            )
+
+        x = np.asarray(x)
```
Comment (mtezzele) [inline src/numanalysislib/basis/chebyshev.py]: keep it close to where you use it (line 65). Otherwise you can do the cast directly in line 65 within the formula

File: src/numanalysislib/basis/chebyshev.py [hunk 5]
```diff
@@ -0,0 +1,187 @@
...
+        except np.linalg.LinAlgError:
+            raise ValueError(
+                "we are having singular matrix, make sure that the nodes are distinct"
+            )
+
+        return coef
+
+    def evaluate(self, coef: np.ndarray, x: np.ndarray) -> np.ndarray:
+        """
+        evaluate the series sum(c_i*T_i(x))
```
Comment (mtezzele) [inline src/numanalysislib/basis/chebyshev.py]: First letter is a capital letter

**Review by mtezzele — COMMENTED** [review]
Please add a test in the Affline test file where you check that building an affinespace with Cheb is working properly.

**Comment by Lemux133** [comment]
> Please add a test in the Affline test file where you check that building an affinespace with Cheb is working properly.

Hi professor, is the following test as expected?

At the end of TestAffinePower class, I could add this test: 

def test_chebyshev_affine(self):
      """
      Test AffinePolynomialBasis works with ChebyshevBasis
      """
      from numanalysislib.basis.chebyshev import ChebyshevBasis
        
      cheb = ChebyshevBasis(degree=2)
      affine = AffinePolynomialBasis(cheb, a=0, b=2)
        
      # physical x=1 mapped to reference x_hat=0, where T_2(0) = -1
      result = affine.evaluate_basis(2, np.array([1.0]))
      np.testing.assert_allclose(result[0], -1.0)
